### ISSUE TITLE
Offline install failure - DNS leaking when node is isolated from upstream DNS

### ISSUE THREAD

**Issue by andrewtappert** [gravitational/gravity#257]
Steps to reproduce:
1. Start with a minimal CentOS install (can be online with Internet access)
2. Edit /etc/resolv.conf to put in a bogus nameserver
(We have also repro'd with other ways of taking the system "offline" besides munging the nameserver, including setting a bogus default gateway, and simply disconnecting the NIC at the VM layer.)
3. Run gravity install from telekube 5.2.5 bundle.
Expecting: successful install.
Finding: it never makes it past site-app installation.

Install output:
```
$ sudo ./gravity install
Fri Feb  8 14:25:51 UTC	Starting installer
Fri Feb  8 14:25:51 UTC	Preparing for installation...
Fri Feb  8 14:26:19 UTC	Installing application telekube:5.2.5
Fri Feb  8 14:26:19 UTC	Starting non-interactive install
Fri Feb  8 14:26:20 UTC	All agents have connected!
Fri Feb  8 14:26:20 UTC	Starting the installation
Fri Feb  8 14:26:21 UTC	Operation has been created
Fri Feb  8 14:26:22 UTC	Execute preflight checks
Fri Feb  8 14:26:24 UTC	Configure packages for all nodes
Fri Feb  8 14:26:28 UTC	Bootstrap master node node-4
Fri Feb  8 14:26:32 UTC	Pull packages on master node node-4
Fri Feb  8 14:27:20 UTC	Install system software on master node node-4
Fri Feb  8 14:27:21 UTC	Install system package teleport:2.4.7 on master node node-4
Fri Feb  8 14:27:23 UTC	Install system package planet:5.2.19-11105 on master node node-4
Fri Feb  8 14:27:47 UTC	Wait for system services to start on all nodes
Fri Feb  8 14:28:28 UTC	Apply labels and taints to Kubernetes nodes
Fri Feb  8 14:28:29 UTC	Label and taint master node node-4
Fri Feb  8 14:28:30 UTC	Bootstrap Kubernetes roles and PSPs
Fri Feb  8 14:28:32 UTC	Populate Docker registry on master node node-4
Fri Feb  8 14:29:08 UTC	Install system applications
Fri Feb  8 14:29:09 UTC	Install system application dns-app:0.1.0
Fri Feb  8 14:29:16 UTC	Install system application bandwagon:5.2.3
Fri Feb  8 14:29:21 UTC	Install system application logging-app:5.0.2
Fri Feb  8 14:29:28 UTC	Install system application monitoring-app:5.2.2
Fri Feb  8 14:29:46 UTC	Install system application tiller-app:5.2.1
Fri Feb  8 14:30:07 UTC	Install system application site:5.2.5
Fri Feb  8 14:37:08 UTC Operation failure: Job has reached the specified backoff limit
Fri Feb  8 14:37:08 UTC	Installation failed in 10m47.907408928s, check /var/log/telekube-install.log for details
---
Installer process will keep running so you can inspect the operation plan using
`gravity plan` command, see what failed and continue plan execution manually
using `gravity install --phase=<phase-id>` command after fixing the problem.
Once no longer needed, this process can be shutdown using Ctrl-C.
```

Tail of telekube-install.log:
```
[ERROR]: failed connecting to https://gravity-site.kube-system.svc.cluster.local:3009/healthz
2019-02-08T14:34:17Z ERRO             "\nERROR REPORT:\nOriginal Error: *url.Error Get https://gravity-site.kube-system.svc.cluster.local:3009/healthz: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)\nStack Trace:\n\t/gopath/src/github.com/gravitational/gravity/tool/gravity/cli/site.go:53 github.com/gravitational/gravity/tool/gravity/cli.statusSite\n\t/gopath/src/github.com/gravitational/gravity/tool/gravity/cli/run.go:224 github.com/gravitational/gravity/tool/gravity/cli.Execute\n\t/gopath/src/github.com/gravitational/gravity/tool/gravity/cli/run.go:79 github.com/gravitational/gravity/tool/gravity/cli.Run\n\t/gopath/src/github.com/gravitational/gravity/tool/gravity/main.go:45 main.run\n\t/gopath/src/github.com/gravitational/gravity/tool/gravity/main.go:36 main.main\n\t/go/src/runtime/proc.go:207 runtime.main\n\t/go/src/runtime/asm_amd64.s:2362 runtime.goexit\nUser Message: failed connecting to https://gravity-site.kube-system.svc.cluster.local:3009/healthz\n" gravity/main.go:36
Container "post-install-hook" changed status from "running" to "terminated, exit code 255".

Container "post-install-hook" changed status from "terminated, exit code 255" to "waiting, reason CrashLoopBackOff".

Container "post-install-hook" restarted, current state is "running".

WARN             syslog not available. reverting to stderr utils/cli.go:82
WARN             syslog not available. reverting to stderr utils/cli.go:82
Fri Feb  8 14:37:07 UTC [ERROR] [node-4] Phase execution failed: Job has reached the specified backoff limit, gravitational.io/site:5.2.5 postInstall hook failed.
```

**Comment by knisbet** [comment]
Thank You again for the report.

I've managed to reproduce this issue, and it appears to be related to the upstream DNS server being unavailable. My reproduction is showing, that we're leaking DNS queries, that should be resolved inside the cluster, to the upstream DNS server, which if unavailable, needs to time out. These queries are happening due to the search entries in resolv.conf, combined with internal queries that are not pinned to DNS root. 

I've been testing a patch that partially addresses the issue, and I believe resolves the installation failure you've been encountering. However, the patch doesn't currently capture every possible DNS query. Most notable, I'm not currently able to pin the docker image references, as a library somewhere see's `leader.telekube.local.:5000/image:tag` as an invalid reference. So that will take some more time to isolate, between gravity and upstream.

Also, I don't think we're in a position to disable copying search paths from the host entirely, as it is a common gravity use case, that although isolated from the internet, internal DNS resolution still works so the cluster software can be integrated with internal resources, without requiring FQDN configuration.

**Comment by kontsevoy** [comment]
@knisbet do we recommend we close this?

**Comment by knisbet** [comment]
@kontsevoy We discussed options around this today. For a couple reasons, it's a significant engineering effort to fully address this issue and also ensure we don't accidentally leak queries in the future. It also seems even in air-gapped installations, we virtually always have working DNS.

The plan is:
- Even though we don't require internet access (air-gapped), for now, we will require the upstream DNS to be functional if configured.
- To make this easier to troubleshoot, we'll ensure satellite tests the upstream DNS as part of pre-installation and runtime health checks. I'll likely copy the coredns health checks (https://coredns.io/plugins/forward/) and do a `. IN NS` query to the upstream.
- I'll also re-evaluate configuring coredns upstream health checks, originally I disabled the coredns health checks, since I wasn't sure if we should health check the customers DNS or not as part of our config, however, after looking at this issue, I think we might have to.


I'm also working on limiting the scope of the patch so that it works with upgrades, but will avoid the hard failure Andrew experienced, which was difficult to debug.
